### PR TITLE
Add Nystromformer blog post

### PR DESCRIPTION
**Pull request by novice03**
Hello, this is a blog post on [Nystromformer](https://arxiv.org/abs/2102.03902). Through this post, I wanted to:

- provide a brief overview of the Nystrom method for approximating matrices
- explain how the Nystrom method can be leveraged to approximate self-attention
- raise awareness of the HuggingFace implementation

Reviewers: @NimaBoscarino

### CODE REVIEW

**Review by NimaBoscarino — COMMENTED** [review]
Thank you so much for writing this!! I left a few comments, but this is way out of my area of expertise so unfortunately I don't think I can provide any useful feedback on the actual content. I think @NielsRogge will be more helpful than me here 😄

**Comment by novice03** [comment]
Thanks a lot for your review @NimaBoscarino. I've applied the suggested changes. Looking forward to suggestions from other reviewers.

File: nystromformer.md [hunk 1]
```diff
@@ -0,0 +1,151 @@
+---
+title: "Nyströmformer: Approximating self-attention in linear time and memory via the Nyström method"
+thumbnail: /blog/assets/86_nystromformer/paper_figure.png
```
[the remark this answers is not in the document]
Comment (NimaBoscarino) [reply nystromformer.md:3]: Totally! Figma is a good choice, or you could use Canva instead.

Here's a ZIP file with a bunch of SVGs that you could choose from for a background, and a HF logo: https://drive.google.com/file/d/1d-uLYKWUV1K10MRmnxXfW6MAUrTd4qaT/view?usp=sharing

And the font we use is "Source Sans Pro", bold.

File: nystromformer.md [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+
+<h1>Nyströmformer: Approximating self-attention in linear time and memory via the Nyström method</h1>
+
+<script async defer src="https://unpkg.com/medium-zoom-element@0/dist/medium-zoom-element.min.js"></script>
+
+## Introduction
+
+Transformers have exhibited remarkable performance on various Natural Language Processing and Computer Vision tasks. Their success can be attributed to the self-attention mechanism, which captures the pairwise interactions between all the tokens in an input. However, the standard self-attention mechanism has a time and memory complexity of \\(O(n^2)\\) (where \\(n\\) is the length of the input sequence), making it expensive to train on long input sequences. 
+
+The [Nyströmformer](https://arxiv.org/abs/2102.03902) is one of many efficient transformer models that approximates standard self-attention with \\(O(n)\\) complexity. Nyströmformer exhibits competitive performance on various downstream NLP tasks while improving upon the efficiency of standard self-attention. The aim of this blog post is to give readers an overview of the Nyström method and how it can be adapted to approximate self-attention.
```
Comment (NielsRogge) [inline nystromformer.md:14]:
```suggestion
The [Nyströmformer](https://arxiv.org/abs/2102.03902) is one of many efficient Transformer models that approximates standard self-attention with \\(O(n)\\) complexity. Nyströmformer exhibits competitive performance on various downstream NLP tasks while improving upon the efficiency of standard self-attention. The aim of this blog post is to give readers an overview of the Nyström method and how it can be adapted to approximate self-attention.
```
I always capitalize Transformer :D btw the Nyströmformer author also showed that it works well on ImageNet, right?

File: nystromformer.md [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+with torch.no_grad():
+    logits = model(**inputs).logits
+
+# retrieve index of [MASK]
+mask_token_index = (inputs.input_ids == tokenizer.mask_token_id)[0].nonzero(as_tuple=True)[0]
+
+predicted_token_id = logits[0, mask_token_index].argmax(axis=-1)
+tokenizer.decode(predicted_token_id)
+```
+Alternatively, we can use pipeline:
```
Comment (NielsRogge) [inline nystromformer.md:148]:
```suggestion
Alternatively, we can use the pipeline API (which handles all the complexity for us):
```

File: nystromformer.md [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+The original implementation of Nyströmformer can be found [here](https://github.com/mlpen/Nystromformer) and the HuggingFace implementation can be found [here](https://github.com/huggingface/transformers/tree/main/src/transformers/models/nystromformer). Let's take a look at a few lines of code (with some comments added) from the HuggingFace implementation. Note that some details such as normalization, attention masking, and depthwise convolution are avoided for simplicity.
+
+```python
+key_layer = self.transpose_for_scores(self.key(hidden_states)) # K
+value_layer = self.transpose_for_scores(self.value(hidden_states)) # V
+query_layer = self.transpose_for_scores(mixed_query_layer) # Q
+
+.
+.
+.
```
Comment (NielsRogge) [inline nystromformer.md:96]: What's the purpose of these 3 dots? :D

File: nystromformer.md [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+
+with torch.no_grad():
+    logits = model(**inputs).logits
+
+# retrieve index of [MASK]
+mask_token_index = (inputs.input_ids == tokenizer.mask_token_id)[0].nonzero(as_tuple=True)[0]
+
+predicted_token_id = logits[0, mask_token_index].argmax(axis=-1)
+tokenizer.decode(predicted_token_id)
+```
```
Comment (NielsRogge) [inline nystromformer.md:147]: Would be nice to include the output here as well, you can check other blog posts to see how the output is added, e.g. https://huggingface.co/blog/how-to-generate

File: nystromformer.md [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+  <medium-zoom background="rgba(0,0,0,.7)" alt="Efficient self-attention with the Nyström method" src="assets/86_nystromformer/paper_figure.png"></medium-zoom>
+  <figcaption>Efficient self-attention with the Nyström method</figcaption>
+</figure>
+
+The three orange matrices above correspond to the three matrices we constructed using the key and query landmarks. Also, notice that there is a DConv box. This corresponds to a skip connection added to the values using a 1D depthwise convolution.
+
+
+## How is Nyströmformer implemented?
+
+The original implementation of Nyströmformer can be found [here](https://github.com/mlpen/Nystromformer) and the HuggingFace implementation can be found [here](https://github.com/huggingface/transformers/tree/main/src/transformers/models/nystromformer). Let's take a look at a few lines of code (with some comments added) from the HuggingFace implementation. Note that some details such as normalization, attention masking, and depthwise convolution are avoided for simplicity.
```
Comment (NielsRogge) [inline nystromformer.md:87]:
```suggestion
The original implementation of Nyströmformer can be found [here](https://github.com/mlpen/Nystromformer) and the HuggingFace implementation can be found [here](https://github.com/huggingface/transformers/blob/main/src/transformers/models/nystromformer/modeling_nystromformer.py). Let's take a look at a few lines of code (with some comments added) from the HuggingFace implementation. Note that some details such as normalization, attention masking, and depthwise convolution are avoided for simplicity.
```

File: nystromformer.md (hunk 2 above)
Comment (novice03) [reply nystromformer.md:14]: Yeah, it looks better capitalized. I shall make that change everywhere. I don't think the author ran ImageNet experiments, but there is an image classification task in the LRA benchmark.

File: nystromformer.md (hunk 4 above)
Comment (novice03) [reply nystromformer.md:96]: I thought they would signify that I left out some part of the code. But, on second thought, they seem unnecessary. I shall remove them.

File: nystromformer.md (hunk 5 above)
Comment (novice03) [reply nystromformer.md:143]: Thanks for the suggestion. Added output for both code blocks.

File: nystromformer.md (hunk 2 above)
Comment (NielsRogge) [reply nystromformer.md]: Yeah I was referring to this: https://github.com/mlpen/Nystromformer/tree/main/ImageNet.

But doesn't seem like he reported any metrics

Comment (NielsRogge) [reply nystromformer.md]: Update: from the README:
> We extended segment-means to compute landmarks without requiring the sequence length divisible by the number of landmarks. Then we used this Nystromformer to perform deployment of T2T-Vit_t-14 for image classification without retraining. Our T2T-ViT-Nys-14 achieves 78% top-1 accuracy, outperforming performer/Linformer +4.3%/+12.7% for the direct deployment.

Comment (novice03) [reply nystromformer.md]: Oh yeah, I forgot about that. I changed 

> various downstream NLP tasks

to

> various downstream NLP and CV tasks

File: nystromformer.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (novice03) [reply nystromformer.md]: Thanks for sharing this. I've pushed the thumbnail I made to this branch. Please let me know what you think.

File: _blog.yml [hunk 7]
```diff
@@ -977,4 +977,13 @@
...
     - nlp
-    - guide
\ No newline at end of file
+    - guide
+
+- local: nystromformer
+  title: "Nyströmformer: Approximating self-attention in linear time and memory via the Nyström method"
+  author: novice03
+  thumbnail: /blog/assets/86_nystromformer/thumbnail.png
+  date: July 11, 2022
```
Comment (osanseviero) [inline _blog.yml:986]: To update before merge

File: nystromformer.md [hunk 8]
```diff
@@ -0,0 +1,189 @@
+---
+title: "Nyströmformer: Approximating self-attention in linear time and memory via the Nyström method"
+thumbnail: /blog/assets/86_nystromformer/thumbnail.png
+---
+
```
Comment (osanseviero) [inline nystromformer.md:5]: It would be great if you could add a data + an author card as in https://github.com/huggingface/blog/blob/main/1b-sentence-embeddings.md?plain=1#L9-L25

**Review by osanseviero — APPROVED** [review]
Thanks for this blog post! I left some comments that are important for the post, but apart from it this looks great 🔥 feel free to re-request review once you think it's ready!

**Comment by novice03** [comment]
Thank you so much for your review @osanseviero. I added the author card and I shall update the date right before this PR is ready to be merged. I also marked this PR as ready for review.

File: _blog.yml [hunk 9]
```diff
@@ -978,7 +978,7 @@
     - sentiment-analysis
     - nlp
     - guide
-
+
```
Comment (osanseviero) [inline _blog.yml:981]: Please remove this minor change

**Review by osanseviero — APPROVED** [review]
This looks great to me! Last week we had a bunch of blog posts, but this week we can merge this. WDYT Of releasing this tomorrow?

**Comment by novice03** [comment]
Sounds good to me! Just to confirm - the date would be August 2nd, right?

**Comment by osanseviero** [comment]
That's right! :partying_face:

**Review by NielsRogge — APPROVED** [review]
LGTM! Just wondering if the code block regarding the implementation adds a lot of value.

File: nystromformer.md [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+<h1>Nyströmformer: Approximating self-attention in linear time and memory via the Nyström method</h1>
+
+<div class="blog-metadata">
+    <small>Published July 11, 2022.</small>
```
Comment (osanseviero) [inline nystromformer.md:9]:
```suggestion
    <small>Published Aug 2, 2022.</small>
```

**Comment by osanseviero** [comment]
Let's ship this :fire: :rocket:

**Comment by osanseviero** [comment]
This is up now! https://huggingface.co/blog

**Comment by novice03** [comment]
Thank you so much!

There seem to be some latex issues, especially in the "How can we adapt the Nyström method to approximate self-attention?" section. How can we fix this?

**Comment by osanseviero** [comment]
Looking into this :)

**Comment by osanseviero** [comment]
Opened https://github.com/huggingface/blog/pull/449

**Comment by julien-c** [comment]
great blogpost @novice03 thanks for contributing!
